lib/src/editorhelper.cpp

```diff
@@ -131,92 +146,207 @@ void resize_buffer(T* src_buffer, T* dst_buffer, int src_w, int src_h, int dst_w
+}
+
+uhdr_resize_effect::uhdr_resize_effect(int width, int height) : m_width{width}, m_height{height} {
+  m_resize_uint8_t = resize_buffer<uint8_t>;
+  m_resize_uint16_t = resize_buffer<uint16_t>;
+  m_resize_uint32_t = resize_buffer<uint32_t>;
+  m_resize_uint64_t = resize_buffer<uint64_t>;
+}
+
+std::unique_ptr<uhdr_raw_image_ext_t> apply_rotate(ultrahdr::uhdr_rotate_effect_t* desc,
+                                                   uhdr_raw_image_t* src) {
   std::unique_ptr<uhdr_raw_image_ext_t> dst;
+  effect_fn_desc_uint8_t rotate_uint8_t = nullptr;
+  effect_fn_desc_uint16_t rotate_uint16_t = nullptr;
+  effect_fn_desc_uint32_t rotate_uint32_t = nullptr;
+  effect_fn_desc_uint64_t rotate_uint64_t = nullptr;
 
-  if (degree == 90 || degree == 270) {
+  if (desc->m_degree == 90) {
+    dst = std::make_unique<uhdr_raw_image_ext_t>(src->fmt, src->cg, src->ct, src->range, src->h,
+                                                 src->w, 1);
+    rotate_uint8_t = desc->m_rotate_90_uint8_t;
+    rotate_uint16_t = desc->m_rotate_90_uint16_t;
+    rotate_uint32_t = desc->m_rotate_90_uint32_t;
+    rotate_uint64_t = desc->m_rotate_90_uint64_t;
+  } else if (desc->m_degree == 270) {
     dst = std::make_unique<uhdr_raw_image_ext_t>(src->fmt, src->cg, src->ct, src->range, src->h,
                                                  src->w, 1);
-  } else if (degree == 180) {
+    rotate_uint8_t = desc->m_rotate_270_uint8_t;
+    rotate_uint16_t = desc->m_rotate_270_uint16_t;
+    rotate_uint32_t = desc->m_rotate_270_uint32_t;
+    rotate_uint64_t = desc->m_rotate_270_uint64_t;
+  } else if (desc->m_degree == 180) {
     dst = std::make_unique<uhdr_raw_image_ext_t>(src->fmt, src->cg, src->ct, src->range, src->w,
                                                  src->h, 1);
+    rotate_uint8_t = desc->m_rotate_180_uint8_t;
+    rotate_uint16_t = desc->m_rotate_180_uint16_t;
+    rotate_uint32_t = desc->m_rotate_180_uint32_t;
+    rotate_uint64_t = desc->m_rotate_180_uint64_t;
   } else {
     return nullptr;
   }
 
   if (src->fmt == UHDR_IMG_FMT_24bppYCbCrP010) {
     uint16_t* src_buffer = static_cast<uint16_t*>(src->planes[UHDR_PLANE_Y]);
     uint16_t* dst_buffer = static_cast<uint16_t*>(dst->planes[UHDR_PLANE_Y]);
-    rotate_buffer_clockwise(src_buffer, dst_buffer, src->w, src->h, src->stride[UHDR_PLANE_Y],
-                            dst->stride[UHDR_PLANE_Y], degree);
+    rotate_uint16_t(src_buffer, dst_buffer, src->w, src->h, -1, -1, src->stride[UHDR_PLANE_Y],
+                    dst->stride[UHDR_PLANE_Y]);
     uint32_t* src_uv_buffer = static_cast<uint32_t*>(src->planes[UHDR_PLANE_UV]);
     uint32_t* dst_uv_buffer = static_cast<uint32_t*>(dst->planes[UHDR_PLANE_UV]);
-    rotate_buffer_clockwise(src_uv_buffer, dst_uv_buffer, src->w / 2, src->h / 2,
-                            src->stride[UHDR_PLANE_UV] / 2, dst->stride[UHDR_PLANE_UV] / 2, degree);
+    rotate_uint32_t(src_uv_buffer, dst_uv_buffer, src->w / 2, src->h / 2, -1, -1,
```

Comment:
This may be wrong, could you double check? Can UV interlaced be rotated the same way as non-interlaced channel?

i.e. if input is

```
u1 v1 u2 v2
u3 v3 u4 v4
```

then after 90 degree rotation it becomes:

```
u3 u1
v3 v1
u4 u2
v4 v2
```

I think the correct way is width should not be divided by 2, and sample every other pixel for U/V channel

Reply:
Dichen, i am treating U and V as one unit and operating on the entire unit. Data type is uint32_t. So both u and v are rotated together

Reply:
Ah OK. Sorry my mistake

Reply:
After rotation, 
u3 v3 u1 v1
u4 v4 u2 v2